packages/core/src/types/ast.ts

```diff
@@ -4,9 +4,14 @@ export interface SourceLoc {
   col: number;
 }
 
+export interface SourceRange {
+  start: SourceLoc;
+  end: SourceLoc;
+}
+
 type Optional<T, K extends keyof T> = Omit<T, K> & Partial<T>;
 
-export type ASTNode = Optional<ConcreteNode, "start" | "end">;
+export type ASTNode = Optional<ConcreteNode, keyof SourceRange>;
```

Comment:
Now I'm seeing the problem after some testing. I think we need to think about how to design the node types again. The current types aren't super useful. Perhaps the right way to go is generics? 
* `ASTNode<T>` contains all the meta properties and data of type `T`.
```ts
interface ASTNode<T extends NodeData> {
  data: T,
  // ...other meta props
}
```
* For convinience, we also need a common super type `NodeData` for all the node data types (`Decl`, `StringLit` etc) so we can do things generically over node data. Seems like the only thing they have in common is `tag`? 
```ts
interface NodeData { 
  tag: string 
}
interface Decl extends NodeData { 
  // ...concrete data here
}
```
* When typing the parsers, we can use `ASTNode<T>` for the output of each rule

Reply:
This is probably a big change. So I guess we could do another PR for this if you think it's a good idea? Lmk if you have any suggestions!

Reply:
Yep this makes a lot of sense to me; it's more or less what I was thinking about doing eventually after mulling over it for a little while. Probably should be done in a separate PR as you suggest (i.e. as a prerequisite to this PR).